Enhance the Stepper widget by adding customizable header and content padding

safwanidrees opened the pull request:
Enhance the Stepper widget by adding customizable header and content padding properties. Default values are set for both paddings to maintain the existing layout behavior.

<!--
Thanks for filing a pull request!
Reviewers are typically assigned within a week of filing a request.
To learn more about code review, see our documentation on Tree Hygiene: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
-->

*Replace this paragraph with a description of what this PR is changing or adding, and why. Consider including before/after screenshots.*

*List which issues are fixed by this PR. You must list at least one issue. An issue is not required if the PR fixes something trivial like a typo.*

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

**Note**: The Flutter team is currently trialing the use of [Gemini Code Assist for GitHub](https://developers.google.com/gemini-code-assist/docs/review-github-code). Comments from the `gemini-code-assist` bot should not be taken as authoritative feedback from the Flutter team. If you find its comments useful you can update your code accordingly, but if you are unsure or disagree with the feedback, please feel free to wait for a Flutter team member's review for guidance on which automated comments should be addressed.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

safwanidrees committed:
Enhance Stepper widget by adding customizable header and content padding properties. Default values are set for both paddings to maintain existing layout behavior.

dkwingsmt commented on packages/flutter/lib/src/material/stepper.dart:
```diff
@@ -379,6 +381,21 @@ class Stepper extends StatefulWidget {
+  /// The padding around the header row in both [StepperType.vertical] and
+  /// [StepperType.horizontal] steppers.
+  ///
+  /// If null, defaults to `EdgeInsets.symmetric(horizontal: 24.0)`.
```
```suggestion
  /// Defaults to `EdgeInsets.symmetric(horizontal: 24.0)`.
```

dkwingsmt reviewed (commented):
Thanks for the contribution. It generally looks good to me. Can you add some unit tests?

safwanidrees commented:
> Thanks for the contribution. It generally looks good to me. Can you add some unit tests?

Sure I can

safwanidrees committed:
Refine Stepper widget documentation and enhance content padding logic. Added tests for customizable header and content padding for both vertical and horizontal steppers, ensuring proper alignment with step icons when margins are applied.

dkwingsmt commented on packages/flutter/test/material/stepper_test.dart:
```diff
@@ -1706,6 +1706,248 @@ void main() {
+      );
+
+      final Stepper stepper = tester.widget<Stepper>(find.byType(Stepper));
+      expect(stepper.headerPadding, customPadding);
```
This test is doing nothing. It constructs an object, then test that the object has the right property, which is trivial and involves nothing about the changed code.

You should instead test that the components are placed at the expected coordinate.

safwanidrees replied:
Updated the tests to validate layout behavior rather than property values. The tests are now validating the position of the displayed header as well as the application of the padding.

safwanidrees committed:
Update Stepper tests to validate padding in layout

dkwingsmt commented on packages/flutter/lib/src/material/stepper.dart:
```diff
@@ -765,6 +782,15 @@ class _StepperState extends State<Stepper> with TickerProviderStateMixin {
     final double? additionalMarginLeft = marginLeft != null ? marginLeft / 2.0 : null;
     final double? additionalMarginRight = marginRight != null ? marginRight / 2.0 : null;
+    // Default content padding for vertical stepper.
+    const EdgeInsetsDirectional defaultContentPadding = EdgeInsetsDirectional.only(
```
Make this a global called `_kDefaultContentPadding

safwanidrees replied:
Done

dkwingsmt commented on packages/flutter/test/material/stepper_test.dart:
```diff
@@ -1706,6 +1706,295 @@ void main() {
+          moreOrLessEquals(8.0));
+    });
+
+    testWidgets('Stepper default headerPadding', (WidgetTester tester) async {
```
Shall we remove this test case, since the previous test cases are sufficient? Testing widget property isn't useful.

safwanidrees replied:
Done

dkwingsmt reviewed (approved):
Generally LGTM except for some minor issues. Thank you!

safwanidrees committed:
Refactor Stepper widget to define default content padding as a constant, improving code clarity and maintainability. This change ensures consistent padding behavior across vertical steppers.

safwanidrees committed:
Add .top position tests for Stepper headerPadding

Verify that headerPadding affects both horizontal (.left) and vertical (.top)
positioning for vertical and horizontal stepper types.

safwanidrees committed:
Refactor headerPadding to use _kDefaultHeaderPadding and getter

Extract default value into constant and create effectiveHeaderPadding getter,
matching the contentPadding pattern. Removes duplication in _buildVerticalHeader
and _buildHorizontal methods.

safwanidrees commented on packages/flutter/lib/src/material/stepper.dart:
```diff
@@ -915,7 +935,7 @@ class _StepperState extends State<Stepper> with TickerProviderStateMixin {
           elevation: widget.elevation ?? 2,
           child: Padding(
-            padding: const EdgeInsets.symmetric(horizontal: 24.0),
+            padding: widget.headerPadding ?? const EdgeInsets.symmetric(horizontal: 24.0),
```
Done! Added `_kDefaultHeaderPadding` constant and `effectiveHeaderPadding` getter, matching the `contentPadding` pattern. Updated both usages to use the getter.

safwanidrees commented on packages/flutter/test/material/stepper_test.dart:
```diff
@@ -1706,6 +1706,295 @@ void main() {
+
+      final Rect customTitleRect = tester.getRect(find.text('Step 1'));
+
+      expect(customTitleRect.left, lessThan(defaultTitleRect.left));
```
Done

safwanidrees committed:
Improve documentation for Stepper widget's contentPadding property by clarifying the default padding behavior for vertical steppers. This change enhances code readability and maintains consistency in padding logic.

safwanidrees commented on packages/flutter/lib/src/material/stepper.dart:
```diff
@@ -379,6 +381,21 @@ class Stepper extends StatefulWidget {
+  /// For [StepperType.horizontal], defaults to `EdgeInsets.all(24.0)`.
+  ///
+  /// For [StepperType.vertical], defaults to
+  /// `EdgeInsetsDirectional.only(start: 60.0, end: 24.0, bottom: 24.0)`. The `start` padding is also increased by the `left` value of [stepIconMargin] if it is provided.
```
Done

safwanidrees committed:
Fix omit_obvious_local_variable_types linter errors

Remove explicit type annotations from const declarations where types can
be inferred.

safwanidrees committed:
Refactor stepper_test.dart to remove explicit type annotation for expected EdgeInsets, allowing type inference. This change aligns with recent linter recommendations for cleaner code.

safwanidrees committed:
Formatted the code properly

safwanidrees commented:
> Also, the analyzer check is failing.

Done @dkwingsmt

safwanidrees committed:
Add _kDefaultHorizontalContentPadding for horizontal stepper fallback

victorsanni commented on packages/flutter/lib/src/material/stepper.dart:
```diff
@@ -123,6 +123,13 @@ const double _kStepSize = 24.0;
 const double _kTriangleSqrt = 0.866025; // sqrt(3.0) / 2.0
 const double _kTriangleHeight = _kStepSize * _kTriangleSqrt;
 const double _kMaxStepSize = 80.0;
+const EdgeInsetsDirectional _kDefaultContentPadding = EdgeInsetsDirectional.only(
```
should this be `_kDefaultVerticalContentPadding`?

victorsanni commented on packages/flutter/lib/src/material/stepper.dart:
```diff
@@ -926,7 +953,7 @@ class _StepperState extends State<Stepper> with TickerProviderStateMixin {
             controller: widget.controller,
             physics: widget.physics,
-            padding: const EdgeInsets.all(24.0),
+            padding: widget.contentPadding ?? _kDefaultHorizontalContentPadding,
```
Maybe we can have both an `effectiveHorizontalContentPadding` and an `effectiveVerticalContentPadding`.

safwanidrees commented on packages/flutter/lib/src/material/stepper.dart:
```diff
@@ -765,6 +793,11 @@ class _StepperState extends State<Stepper> with TickerProviderStateMixin {
     final double? additionalMarginLeft = marginLeft != null ? marginLeft / 2.0 : null;
     final double? additionalMarginRight = marginRight != null ? marginRight / 2.0 : null;
+    // Adjust padding to align content with step icon when stepIconMargin is set.
+    final EdgeInsetsGeometry effectiveContentPadding =
```
Vertical

safwanidrees committed:
Rename default content padding constant to _kDefaultVerticalContentPadding and update references in the Stepper widget to improve clarity and maintain consistency in vertical padding logic.

safwanidrees committed:
Refactor Stepper widget to use a local variable for effective horizontal content padding, enhancing code clarity and maintainability. This change ensures consistent padding behavior in the horizontal layout.

dkwingsmt reviewed (approved):
Thank you for your contribution!

dkwingsmt committed:
Merge branch 'master' into stepper-customizable-padding
